mcs/class/corlib/Mono/SafeStringMarshal.cs

```diff
@@ -0,0 +1,46 @@
+//
+// Safe wrapper for a string and its UTF8 encoding
+//
+// Authors:
+//   Aleksey Kliger <aleksey@xamarin.com>
+//   Rodrigo Kumpera <kumpera@xamarin.com>
+//
+// Copyright 2016 Dot net foundation.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System;
+using System.Runtime.CompilerServices;
+
+namespace Mono  {
+	internal sealed class SafeStringMarshal : IDisposable {
```

Comment:
In general, we always side with not allocating unless we absolutely have to.

C# performance depends a lot of how fast we produce garbage